src/torchmetrics/functional/classification/confusion_matrix.py

```diff
@@ -11,15 +11,335 @@
+        if preds.shape[1] != num_classes:
+            raise ValueError(
+                "If `preds` have one dimension more than `target`, `preds.shape[1]` should be"
+                " equal to number of classes."
+            )
+        if preds.shape[2:] != target.shape[1:]:
+            raise ValueError(
+                "If `preds` have one dimension more than `target`, the shape of `preds` should be"
+                " (N, C, ...), and the shape of `target` should be (N, ...)."
+            )
+    elif preds.ndim == target.ndim:
+        if preds.shape != target.shape:
+            raise ValueError(
+                "The `preds` and `target` should have the same shape,",
+                f" got `preds` with shape={preds.shape} and `target` with shape={target.shape}.",
+            )
+    else:
+        raise ValueError(
+            "Either `preds` and `target` both should have the (same) shape (N, ...), or `target` should be (N, ...)"
+            " and `preds` should be (N, C, ...)."
+        )
+
+    unique_values = torch.unique(target)
+    if ignore_index is None:
+        check = len(unique_values) > num_classes
+    else:
+        check = len(unique_values) > num_classes + 1
+    if check:
+        raise RuntimeError(
+            "Detected more unique values in `target` than `num_classes`. Expected only "
+            f"{num_classes if ignore_index is None else num_classes + 1} but found"
+            f"{len(unique_values)} in `target`."
+        )
+
+    if not preds.is_floating_point():
+        unique_values = torch.unique(preds)
+        if len(unique_values) > num_classes:
+            raise RuntimeError(
+                "Detected more unique values in `preds` than `num_classes`. Expected only "
+                f"{num_classes} but found {len(unique_values)} in `preds`."
+            )
+
+
+def _multiclass_confusion_matrix_format(
+    preds: Tensor, target: Tensor, ignore_index: Optional[int] = None
+) -> Tuple[Tensor, Tensor]:
+    # Apply argmax if we have one more dimension
+    if preds.ndim == target.ndim + 1:
+        preds = preds.argmax(dim=1)
+
+    preds = preds.flatten()
+    target = target.flatten()
+
+    if ignore_index is not None:
+        idx = target != ignore_index
+        preds = preds[idx]
+        target = target[idx]
+
+    return preds, target
+
+
+def _multiclass_confusion_matrix_update(preds: Tensor, target: Tensor, num_classes: int) -> Tensor:
+    unique_mapping = (target * num_classes + preds).to(torch.long)
+    bins = _bincount(unique_mapping, minlength=num_classes**2)
+    return bins.reshape(num_classes, num_classes)
+
+
+def _multiclass_confusion_matrix_compute(confmat: Tensor, normalize: Optional[str] = None) -> Tensor:
+    return _confusion_matrix_reduce(confmat, normalize, multilabel=False)
+
+
+def multiclass_confusion_matrix(
+    preds: Tensor,
+    target: Tensor,
+    num_classes: int,
+    ignore_index: Optional[int] = None,
+    normalize: Optional[str] = None,
```

Comment:
nitpick: Again would be a bit more verbose :]
```suggestion
    normalize: Optional[Literal["true", "pred", "all", "none"]] = None,
```